### ISSUE TITLE
Quantal response equilibrium (QRE) as a solver method

### ISSUE THREAD

**Issue by sbenthall** [scikit-agent/scikit-agent#353]
## Summary

Add quantal response equilibrium (QRE) — the standard "soft"/boundedly-rational solution concept for games — as a new solver **method**, complementing `NeuralBestResponse`, `ExactBestResponse`, and `TabularBestResponseSolver`. Unlike those, a QRE method's `best_response` returns a *stochastic* rule: rather than the action that maximizes payoff, each action is played with probability proportional to `exp(λ · expected_payoff)`, for a rationality parameter λ ≥ 0 (λ→0: uniform random; λ→∞: recovers the corresponding exact-best-response/Nash behavior). This is currently the one missing solution concept in the strategic-equilibrium roadmap line, and it's the concept the Prisoners Dilemma / Tree Killer MACID models are working around today via a continuous `[0, 1]` relaxation.

## Motivation

- The existing `solve_in_relevance_order` cyclic-component handling and `solve_symmetric_equilibrium` both target **pure**-strategy fixed points and document this as a current limit ("Mixed-strategy equilibria and recurring cyclic games are not supported by this schedule").
- QRE gives every action strictly positive probability for any finite λ, which is what makes a solved game usable later as an *observation model* (a prerequisite for structural estimation / inverse game theory — see "Explicitly out of scope" below, tracked separately).
- QRE fixed points are guaranteed to exist for any finite λ by Brouwer's theorem (the logit response map is continuous, unlike argmax), which is a strictly easier existence question than Nash's.

## Dependencies

- **#352 (discrete action space controls)** — QRE needs a finite, enumerable action set to define a categorical distribution over. This is a hard prerequisite for anything beyond the binary case.
- **#341 / #347 (randomizing controls)** — still under design discussion. The shape a QRE method's output takes depends on how this resolves:
  - If a randomizing control's shock ends up as an ordinary parent in the control's information set (per the proposal raised in #347's review), a QRE method just returns an ordinary deterministic rule over the augmented information set (inverse-CDF/threshold sampling from the logit distribution) — no new "rule" type needed anywhere else in the solver stack.
  - If declared probabilities remain a distinct first-class thing (closer to #347 as currently written), a QRE method returns that probability object directly, and downstream code (schedules, `rule_distance`, simulation) needs to know how to handle it.
  - **This issue should not be started until that design question is resolved**, since it determines the method's return type. Worth raising QRE explicitly in that discussion as a concrete test case.
  - Separately: #347's randomization mechanism is Bernoulli-only (one uniform shock, one threshold). QRE for a >2-action game needs a categorical generalization (e.g., Gumbel-max, or a multi-threshold inverse-CDF over the action set) that doesn't exist yet in either design. Worth scoping as a small sub-task here, or spinning out as its own issue if #341/#347's resolution doesn't already generalize past binary.

## Proposed design

Two candidate implementations, not mutually exclusive:

**(a) An explicit `QuantalResponse` method**, sitting next to `TabularBestResponseSolver` (which already assumes "candidate actions" and "a tabulated payoff table" — the natural home for this). Computes the softmax of expected payoffs over the candidate action set directly.

**(b) An entropy-regularized `NeuralBestResponse` variant.** Adding a KL/entropy penalty to the training objective is the neural/maximum-entropy-RL analogue of logit QRE, with the regularization weight playing the role of `1/λ`. This may fit more naturally where the design conversation in #347 is heading (any solver, including a neural one, learning a policy over an augmented information set that includes random inputs).

Recommend building (a) first as the reference implementation (it's simple, exactly matches the classical definition, and is easy to test against closed-form 2×2 cases), and treating (b) as a follow-on once the shock-in-information-set design settles.

## Two solution concepts to support

**1. Simultaneous / normal-form case** (the cyclic-relevance-component case — e.g., Prisoners Dilemma). Standard logit QRE: fixed-point iteration over the joint mixed-strategy profile.

**2. Sequential / extensive-form case** (the acyclic-relevance-component case, and MACIDs generally — e.g., Tree Killer). This needs **Agent QRE (AQRE)** (McKelvey & Palfrey 1998): each information set softmaxes over *continuation values*, computed via a stochastic backward pass rather than deterministic backward induction. This is the natural generalization of what `solve_in_relevance_order`'s acyclic pass already does, and should probably live as an option on that pass rather than a wholly separate schedule.

## Pseudocode

**(1) Simultaneous case — logit QRE via damped fixed-point iteration**, in the style of the existing `solve_symmetric_equilibrium`:

```python
def solve_qre_simultaneous(payoff_fn, action_sets, lam, damping=0.5,
                            tol=1e-8, max_iterations=1000):
    # payoff_fn(i, a_i, sigma_minus_i) -> expected payoff to player i of
    #     playing action a_i, given the other players' current mixed
    #     strategies sigma_minus_i
    # sigma[i] is a probability vector over action_sets[i]

    sigma = {i: uniform(action_sets[i]) for i in players}

    for iteration in range(max_iterations):
        sigma_new = {}
        for i in players:
            payoffs = [payoff_fn(i, a, sigma_minus(sigma, i))
                       for a in action_sets[i]]
            # logit response
            sigma_new[i] = softmax(lam * payoffs)

        # damped update, mirrors solve_symmetric_equilibrium
        sigma_next = {
            i: (1 - damping) * sigma[i] + damping * sigma_new[i]
            for i in players
        }

        if max(total_variation(sigma_next[i], sigma[i]) for i in players) < tol:
            return sigma_next, {"converged": True, "iterations": iteration}

        sigma = sigma_next

    return sigma, {"converged": False, "iterations": max_iterations}
```

Notes:
- `damping` is not optional here either, for the same reason as Cournot: the logit best-response map need not be a contraction for large λ or many players, and undamped iteration can cycle.
- `total_variation` (or KL) replaces the scalar `rule_distance` used for point-valued rules — this is the piece that needs the design question above resolved, since it determines what a "rule" is being compared.
- λ = 0 should short-circuit to the exact uniform distribution (avoid a wasted/degenerate iteration).

**(2) Sequential case — Agent QRE via a stochastic backward pass**, generalizing `solve_in_relevance_order`'s acyclic handling:

```python
def solve_aqre_sequential(macid, lam):
    # visits decision/utility nodes in topological (relevance) order,
    # last-to-first, exactly where backward induction currently visits
    # them for the acyclic case

    continuation_value = {}  # node -> expected value, keyed by parent-state

    for node in reversed(topological_order(macid)):
        if is_terminal(node):
            continuation_value[node] = payoff(node)

        elif is_decision(node):
            iset = information_set(node)
            for state in enumerate_states(iset):
                payoffs = [
                    expected_value(node, a, state, continuation_value)
                    for a in action_set(node)
                ]
                # store the *distribution*, not the argmax
                policy[node][state] = softmax(lam * payoffs)

            continuation_value[node] = {
                state: sum(policy[node][state][a] *
                           expected_value(node, a, state, continuation_value)
                           for a in action_set(node))
                for state in enumerate_states(iset)
            }

        elif is_chance(node):
            continuation_value[node] = expectation_over(node, continuation_value)

    return policy
```

Notes:
- The key difference from plain backward induction: `continuation_value` at a decision node is the *softmax-weighted average* over actions, not the max. This is what makes the whole pass "stochastic" rather than deterministic, and is the direct generalization of the tree-based `eq_path` idea from the Dewhurst post (linked below) to MACID's relevance-graph solving.
- λ→∞ recovers standard backward induction (`softmax` → `argmax`).
- This should reuse the acyclic branch of `solve_in_relevance_order` as much as possible — the traversal order is identical, only the per-node update rule changes.

## Acceptance criteria

- A `QuantalResponse` method (or equivalent) exists satisfying the three required method operations (`best_response`, `rule_distance`, `initial_policies`), for finite/discrete action sets.
- λ = 0 gives (approximately) uniform random play; large λ approaches the profile found by `ExactBestResponse` / `TabularBestResponseSolver` on the same game.
- Simultaneous case validated against a known closed-form 2×2 logit QRE (e.g. a standard experimental-game-theory example) and against the un-relaxed, discretized Prisoners Dilemma block.
- Sequential/AQRE case validated against the Tree Killer MACID, discretized, checked against backward induction in the λ→∞ limit.
- Damping behavior documented and tested the same way as `solve_symmetric_equilibrium`'s (report `converged: False` rather than a silently wrong iterate).
- Works with `solve_in_relevance_order` for both the acyclic (AQRE) and cyclic (simultaneous logit QRE) cases, or a documented reason why one of those needs a separate entry point.

## Explicitly out of scope (tracked separately)

- Computing the *likelihood* of an observed play against a solved QRE profile, for use as an observation/conditioning model. (Comes close to free for the simultaneous case once this lands — it's a direct lookup — but is a separate piece of work, and remains a real gap for pure-strategy/non-QRE solves.)
- Structural estimation / inverse game theory (inferring latent payoff or calibration parameters from observed play) — downstream of the likelihood item above.
- Homotopy/path-following solvers (Turocy 2005/2010, as implemented in Gambit) as a more numerically robust alternative to fixed-point iteration — worth a follow-up issue if fixed-point iteration proves too fragile in practice, but out of scope for a first implementation.

## References

- McKelvey, R. D., & Palfrey, T. R. (1995). Quantal response equilibria for normal form games. *Games and Economic Behavior*, 10(1), 6–38.
- McKelvey, R. D., & Palfrey, T. R. (1998). Quantal response equilibria for extensive form games. *Experimental Economics*, 1(1), 9–41.
- Turocy, T. L. (2005). A dynamic homotopy interpretation of the logistic quantal response equilibrium correspondence. *Games and Economic Behavior*, 51(2), 243–263.
- Gambit project docs on `gambit-logit` / `pygambit.qre` (reference implementation of the homotopy method, and of maximum-likelihood fitting of λ from observed play).

## Note on authorship

This issue was drafted by Claude (Anthropic), at the request of a scikit-agent maintainer, based on a conversation working through how ideas from David Dewhurst's post [["non-cooperative game theory inside record-based probabilistic programs made 'easy'"](https://daviddewhurst.xyz/econ/tools/game-theory-ppl.html)](https://daviddewhurst.xyz/econ/tools/game-theory-ppl.html) map onto scikit-agent's existing architecture (Blocks, MACIDs, the method/schedule solver split) and current open issues (#341, #347, #352). The pseudocode and design tradeoffs above have not been implemented or tested against the codebase, and the references should be checked against the actual papers before being treated as authoritative. A human should review this for fit with the project's conventions and current design discussions before it's treated as a committed plan.
